[RDY] vim-patch:8.0.{805,806,810,1012,1017}

janlazo opened the pull request:
**vim-patch:8.0.0805: GUI test fails with gnome2**

Problem:    GUI test fails with gnome2.
Solution:   Set $HOME to an existing directory.
vim/vim@d1ee004

**vim-patch:8.0.0806: tests may try to create XfakeHOME twice**

Problem:    Tests may try to create XfakeHOME twice.
Solution:   Avoid loading setup.vim twice.
vim/vim@f98246d

**vim-patch:8.0.0810: MS-Windows: tests still hang**

Problem:    MS-Windows: tests still hang.
Solution:   Only create the XfakeHOME directory if it does not exist yet.
vim/vim@d0b6c6c

**vim-patch:8.0.1012: MS-Windows: problem with $HOME when is was set internally**

Problem:    MS-Windows: Problem with $HOME when is was set internally.
Solution:   Only use the $HOME default internally. (Yasuhiro Matsumoto, closes
            vim/vim#2013)
vim/vim@48340b6

**vim-patch:8.0.1017: test for MS-Windows $HOME always passes**

Problem:    Test for MS-Windows $HOME always passes.
Solution:   Rename the test function.  Make the test pass.
vim/vim@dde6034

janlazo committed:
vim-patch:8.0.1012: MS-Windows: problem with $HOME when is was set internally

Problem:    MS-Windows: Problem with $HOME when is was set internally.
Solution:   Only use the $HOME default internally. (Yasuhiro Matsumoto, closes
            vim/vim#2013)
https://github.com/vim/vim/commit/48340b62e812dc9280f621a2eb6db76d43555c66

Restore vim_getenv() behaviour for $HOME on Windows.

janlazo committed:
vim-patch:8.0.1017: test for MS-Windows $HOME always passes

Problem:    Test for MS-Windows $HOME always passes.
Solution:   Rename the test function.  Make the test pass.
https://github.com/vim/vim/commit/dde60341115effc9789742c2f4565d184e98b6b5

Assume $HOME contains the literal absolute path.
Nested environment variable expansion is unsupported.

justinmk commented on src/nvim/testdir/Makefile:
```diff
@@ -115,7 +115,8 @@ NEW_TESTS ?= \
 	    test_alot_utf8.res \
-	    test_alot.res
+	    test_alot.res \
+	    test_windows_home.res
```
should go after test_writefile

justinmk commented:
Looks like the test fails in appveyor:

```
[00:16:32] From test_windows_home.vim:
[00:16:32] Executing Test_WindowsHome()
[00:16:32] Executed 1 test
[00:16:32] 1 FAILED:
[00:16:32] Found errors in Test_WindowsHome():
[00:16:32] function RunTheTest[35]..Test_WindowsHome[14]..CheckHome line 1: Expected 'C:\\foobar' but got ''
[00:16:32] function RunTheTest[35]..Test_WindowsHome[19]..CheckHome line 1: Expected 'C:\\' but got ''
[00:16:32] function RunTheTest[35]..Test_WindowsHome[25]..CheckHome line 1: Expected 'C:\\foo' but got ''
[00:16:32] function RunTheTest[35]..Test_WindowsHome[39]..CheckHome line 1: Expected 'C:\\foo\\bar' but got '%USERPROFILE%\\bar'
[00:16:32] function RunTheTest[35]..Test_WindowsHome[39]..CheckHome line 2: Expected 'C:\\foo\\bar' but got '%USERPROFILE%\\bar'
```

Maybe it's worth converting

justinmk commented:
Side-note: to get old tests visible on appveyor, we could mark the currently-failing ones as "skip".  Then new tests (like in this PR) will be visible if they fail.

janlazo commented:
Should those tests pass since `$HOME` should be internal?

justinmk commented:
Not sure. Do the test steps work if you try them manually? 

The code looks fine to me, we are using the `homedir` static var ...

janlazo commented:
Tests still fail on Appveyor. Leaving this as blocked for now.

janlazo commented:
May need https://github.com/vim/vim/commit/137374fd6538cf9dee0cb22907728d8fdecb5832 to unlet environment variables. Not sure if it works on Windows.

Is `$HOME` set in Appveyor prior to running an nvim instance?

justinmk commented:
> Is $HOME set in Appveyor prior to running an nvim instance

It's not listed in https://www.appveyor.com/docs/environment-variables/ and we don't set it in `cmake/RunTests.cmake` nor `build.ps1`. But it is set in the Vim oldtests, in `src/nvim/testdir/setup.vim` (which doesn't apply to the Nvim lua tests).

janlazo commented:
Would migrating the new-style tests to a `test/functional/legacy/windows_home_spec.lua` help?

janlazo commented:
Can't work on this until https://github.com/neovim/neovim/pull/8643 is resolved and there is way for nvim to `:unlet`  environment variables.

justinmk commented:
@janlazo there was a recent vim-patch to `:unlet` env vars.

janlazo commented:
If you're referring to vim-patch:8.0.1832, I haven't tried it and I'd rather work on it  after #8643. There are Coverity and ASAN patches that I want to try out first.

Is there a design difference between Vim and Neovim for environment variables? Vim has the `var == NULL || *var == NUL` check to treat empty/unset variables the same. Was that removed or was there a recent patch that introduced it, prior to 8.0.1832?

janlazo commented on src/nvim/testdir/test_windows_home.vim:
```diff
@@ -0,0 +1,121 @@
+    let $HOMEPATH = 'unused'
+    call CheckHome('C:\bar', 'C:\bar')
+
+    " If $HOME contains %USERPROFILE% it is expanded
```
I don't mind if this case is not supported in nvim. I don't set `HOME` to expand an environment variable.

justinmk replied:
Yeah that's a weird workaround. We intentionally don't recursively expand env vars, IIRC. cc @ZyX-I

justinmk commented:
> Is there a design difference between Vim and Neovim for environment variables? Vim has the `var == NULL || *var == NUL` check to treat empty/unset variables the same.

It very much depends on the context. We do check for empty in some cases.

janlazo commented on src/nvim/os/env.c:
```diff
@@ -613,6 +613,12 @@ char *vim_getenv(const char *name)
 
+#ifdef WIN32
+  if (STRCMP(name, "HOME") == 0) {
+    return homedir;
```
I restored the behavior to match vim (replace `WIN32` with `MSWIN`). I don't know how the `let $HOME = '%USERPROFILE%'` test passed.

janlazo committed:
vim-patch:8.0.0805: GUI test fails with gnome2

Problem:    GUI test fails with gnome2.
Solution:   Set $HOME to an existing directory.
https://github.com/vim/vim/commit/d1ee0043c0360f6b504c0283edd2a844ef73a349

janlazo committed:
vim-patch:8.0.0806: tests may try to create XfakeHOME twice

Problem:    Tests may try to create XfakeHOME twice.
Solution:   Avoid loading setup.vim twice.
https://github.com/vim/vim/commit/f98246d4849e670c6660e97887428dcddf88dc9f

janlazo committed:
vim-patch:8.0.0810: MS-Windows: tests still hang

Problem:    MS-Windows: tests still hang.
Solution:   Only create the XfakeHOME directory if it does not exist yet.
https://github.com/vim/vim/commit/d0b6c6c54e0f3c2984e1d9b316a587ef14cfd53a

janlazo commented:
It should be fine now (forgot to copy `homedir`). Not sure why `homepath` must have a default value since `homedrive` and `homepath` should be both defined but it's required for a test to pass. `let $HOME = '%USERPROFILE%\bar'` test is skipped because I don't have use-case for it. I'd rather use batchfiles  than tinker with the Control Panel.

justinmk commented:
So just to be clear, the [vim behavior](https://github.com/neovim/neovim/pull/8485#discussion_r204226646) is _not_ restored, right?

janlazo commented:
@justinmk I restored Vim's special handling of unset/empty `HOME`. Nested env var expansion is unsupported. Previous builds didnt run the test when i made that comment because the nvim instance crashed after homedir was freed (forgot to return output of xstrdup for `HOME`).

janlazo commented:
I didn't port the following code for nested environment variable expansion because it's not part of 8.0.1012 and is wrong; (what about multiple env vars?)

```
    /*
      * Weird but true: $HOME may contain an indirect reference to another
      * variable, esp. "%USERPROFILE%".  Happens when $USERPROFILE isn't set
      * when $HOME is being set.
      */
     if (var != NULL && *var == '%')
     {
 	char_u	*p;
 	char_u	*exp;
 
 	p = vim_strchr(var + 1, '%');
 	if (p != NULL)
 	{
 	    vim_strncpy(NameBuff, var + 1, p - (var + 1));
 	    exp = mch_getenv(NameBuff);
 	    if (exp != NULL && *exp != NUL
 					&& STRLEN(exp) + STRLEN(p) < MAXPATHL)
 	    {
 		vim_snprintf((char *)NameBuff, MAXPATHL, "%s%s", exp, p + 1);
 		var = NameBuff;
 	    }
 	}
     }
```

jamessan commented:
> I forgot if ! is also legal but it's also used for environment variables.

Where?  I've never seen that syntax and it doesn't work in `cmd.exe`.

janlazo commented:
@jamessan I use it for relative filepaths on fzf. See https://github.com/junegunn/fzf/commit/7f0caf06833130727b238a2afeb58f3b5096a34c#diff-4425af3239f24c731114661daef5a811R62

jamessan commented:
Ah, that's specific to using the `/v:on` flag when running `cmd.exe` and not something generally supported.
